github-runner update to 2.286.0 with dotnet sdk 6.0

kfollesdal opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
GitHub have released new github-runner version 2.286.0. 

https://github.com/actions/runner/releases/tag/v2.286.0

Where dotnet is bumped from 3 to 6. I have updated version, sha and dotnet version. But need some help to update dependencies in deps.nix. How to do we update deps.nix automatically? 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

kfollesdal committed:
github-runne: 2.285.1 -> 2.286.0

veehaitch commented:
> But need some help to update dependencies in deps.nix. How to do we update deps.nix automatically?

We currently don't have an automatized way to do that but any help to achieve an automated solution would be highly appreciated 🙂 
Also see: https://github.com/NixOS/nixpkgs/pull/141915#discussion_r730381197

kfollesdal commented:
Have now updated deps.nix. But postConfigure fails. `crossgen` do not longer exist in this location. Have looked around but can not find it. Do somebody with dotnet knowledge know how to slow this. Have 'crossgen' change name or got replaced?
```
  postConfigure = ''
    # `crossgen` dependency is called during build
    patchelf \
      --set-interpreter "$(cat $NIX_CC/nix-support/dynamic-linker)" \
      --set-rpath "${lib.makeLibraryPath [ stdenv.cc.cc.lib ]}" \
      $HOME/.nuget/packages/microsoft.netcore.app.runtime.${runtimeId}/*/tools/crossgen
  '';
```

veehaitch commented:
Just remove the `postConfigure` hook. Also needs to ignore some new online tests: https://github.com/veehaitch/nixpkgs/commit/0c050c1ab7df76ad47514b53e800669a3e3baba5

Afterward, building works and tests pass.

kfollesdal commented:
Thanks @veehaitch. `github-runner` version 2.286.0 now also builds and work for me also. Fun and confidence boost that I found the exactly same fix  as you before I read your comment. Made my day :-) 

Shall I squash the commits into one: `github-runne: 2.285.1 -> 2.286.0`. Or keep commits as is?

kfollesdal commented:
Added modified version of `create_deps.sh ` from [Wasabibackend](https://github.com/NixOS/nixpkgs/blob/ff9d9c338734708d3b421516cd0f8618eb3f1ace/pkgs/applications/blockchains/wasabibackend/create_deps.sh) that I use to make deps.nix. 

Agree with you that it is a mess but probably nice to have for next time.

veehaitch commented:
> Shall I squash the commits into one: `github-runne: 2.285.1 -> 2.286.0`. Or keep commits as is?

Personally, I would reduce it to two commits, one for the version update changes and one for the update script.

Going to test the PR later today but looks pretty good.

veehaitch commented:
Result of `nixpkgs-review pr 152387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>github-runner</li>
  </ul>
</details>

kfollesdal commented:
Try to simplify the create_deps script.

kfollesdal committed:
github-runner: Add script to create deps.nix.

Add create-deps-file script at github-runner.passthru.createDepsFile
Recreate deps.nix with new script.

kfollesdal commented:
@veehaitch. After inspiration from [pkgs/build-support/build-dotnet-module/default.nix](https://github.com/NixOS/nixpkgs/blob/eaeea6e9337b93d2acb7c2aa27e3f4617408375a/pkgs/build-support/build-dotnet-module/default.nix#L120) made a simpler script to create `deps.nix`. Added it at `github-runner.passthru.createDepsFile`

Run it with `$(nix-build -A github-runner.passthru.createDepsFile)/bin/create-deps-file`, the you can copy result file into github-runner folder.

kfollesdal commented:
Result of `nixpkgs-review pr 152387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>github-runner</li>
  </ul>
</details>

veehaitch commented on pkgs/development/tools/continuous-integration/github-runner/default.nix:
```diff
@@ -276,6 +273,34 @@ stdenv.mkDerivation rec {
+      printf "\n* Restore ${pname} dotnet project\n"
+      dotnet restore src/ActionsRunner.sln --packages nuget_pkgs --no-cache --force --runtime linux-x64
+
+      printf "\n* Make deps.nix file\n"
+      deps_file="/tmp/${pname}-deps.nix"
```
Maybe also allow passing the output filename as an argument to the script:

```suggestion
      deps_file=''${1-"/tmp/${pname}-deps.nix"}
      printf "\n* Make %s file\n" "$(basename "$deps_file")"
```

That way we could directly overwrite `deps.nix`:

```bash
$(nix-build -A github-runner.passthru.createDepsFile)/bin/create-deps-file \
  ./pkgs/development/tools/continuous-integration/github-runner/deps.nix
```

kfollesdal replied:
Very good idea.

veehaitch commented on pkgs/development/tools/continuous-integration/github-runner/default.nix:
```diff
@@ -276,6 +273,34 @@ stdenv.mkDerivation rec {
+      deps_file="/tmp/${pname}-deps.nix"
+      nuget-to-nix nuget_pkgs > "$deps_file"
+      printf "\n* Dependency file writen to %s" "$deps_file"
+     '';
```
Minor indention nit:

```suggestion
    '';
```

veehaitch commented on pkgs/development/tools/continuous-integration/github-runner/default.nix:
```diff
@@ -276,6 +273,34 @@ stdenv.mkDerivation rec {
+    runtimeInputs = [ dotnetSdk nuget-to-nix ];
+    text = ''
+      printf "\n* Setup workdir\n"
+      workdir="$(mktemp -d /tmp/github-runner.XXX)"
```
Just for consistency:

```suggestion
      workdir="$(mktemp -d /tmp/${pname}.XXX)"
```

veehaitch reviewed (commented):
Very nice update! Tested the runner on `x86_64-linux`; works like a charm. Also the update script works great.

veehaitch commented:
Having the PR tested on `aarch64-linux` would also be great (maybe @newAM).

Also, I'd very much welcome you as an additional maintainer, if you don't mind, @kfollesdal 🙂

kfollesdal committed:
github-runner: Add @kfollesdal to maintainers

kfollesdal commented:
Thanks @veehaitch for corrections and suggestion. I have commit and squash them. But had to make a little change on script to get cli args to work: https://github.com/NixOS/nixpkgs/pull/152387/commits/a1bc9835b8c760c310dc62be93684904391abcb4

Thanks for wanting me as maintainer :blush: Have added myself to list.

veehaitch committed:
github-runner: Extend script to update all platform deps

Runs `dotnet restore` for all supported platforms (`x86_64-linux` and
`arm64-linux` at the moment) in the update script.

veehaitch commented:
I updated the script to fetch the dependencies of all supported systems: https://github.com/veehaitch/nixpkgs/commit/8ccf49fb2705c2f03e7f43a45ccf370b8ace8e9a

kfollesdal commented:
Result of `nixpkgs-review pr 152387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>github-runner</li>
  </ul>
</details>

kfollesdal commented:
> I updated the script to fetch the dependencies of all supported systems: [veehaitch@8ccf49f](https://github.com/veehaitch/nixpkgs/commit/8ccf49fb2705c2f03e7f43a45ccf370b8ace8e9a)

Very nice, I cherry picked it into to this PR. Ok? Or do you want it in a separate PR.

kfollesdal commented:
Result of `nixpkgs-review pr 152387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>github-runner</li>
  </ul>
</details>

kfollesdal commented:
Result of `nixpkgs-review pr 152387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>github-runner</li>
  </ul>
</details>

kfollesdal commented:
Result of `nixpkgs-review pr 152387` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>github-runner</li>
  </ul>
</details>

kfollesdal commented:
@veehaitch have now tested with new updates, works fine. If you do not have any correction the PR is ready to merg. Will be offline until sunday/monday.

veehaitch reviewed (approved):
Thanks for the effort, the derivation is a lot more maintainable now. 

LGTM

newAM reviewed (approved):
Works on aarch64-linux.

prusnak commented:
Have a look at `buildDotnetModule` - using this would simplify lot of the stuff. Example of usage: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/blockchains/btcpayserver/default.nix

prusnak commented:
Have a look at `buildDotnetModule` - using this would simplify lot of the stuff. Example of usage: https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/blockchains/btcpayserver/default.nix

Using that would allow you to get rid of `runtimeIds`, `nugetSource`, `createDepsFile`, etc.

// cc @IvarWithoutBones

IvarWithoutBones commented:
Based on a quick look at this derivation, the only functionality missing from `buildDotnetModule` would be building unit tests seperately from the main package, with different flags: https://github.com/NixOS/nixpkgs/blob/8b138c90318712d0a0c19732a32a4d1394f5168f/pkgs/development/tools/continuous-integration/github-runner/default.nix#L198

Currently it's compiled together with the installed package, with shared flags. I'll look into implementing that functionality, in the meantime I think we can merge this.

 I'm not sure we should build unit tests in the `checkPhase` btw, shouldn't that be done in `buildPhase`?

> Using that would allow you to get rid of `runtimeIds`, `nugetSource`, `createDepsFile`, etc.

Also, runtime IDs are not yet supported by `buildDotnetModule`. Though the implementation used now could just be copied over.

prusnak commented:
Thanks for the review @IvarWithoutBones!

Let's merge as is then and focus on improvements in the following PRs.
